{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "24956209",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import chain\n",
    "\n",
    "# Hi-C utilities imports:\n",
    "import cooler\n",
    "import bioframe\n",
    "import cooltools\n",
    "from cooltools.lib.numutils import fill_diag\n",
    "from cooltools import insulation\n",
    "from packaging import version\n",
    "if version.parse(cooltools.__version__) < version.parse('0.5.2'):\n",
    "    raise AssertionError(\"tutorials rely on cooltools version 0.5.2 or higher,\"+\n",
    "                         \"please check your cooltools version and update to the latest\")\n",
    "\n",
    "# Visualization imports:\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import LogNorm\n",
    "import matplotlib.patches as patches\n",
    "from matplotlib.ticker import EngFormatter\n",
    "\n",
    "# helper functions for plotting\n",
    "bp_formatter = EngFormatter('b')\n",
    "def format_ticks(ax, x=True, y=True, rotate=True):\n",
    "    \"\"\"format ticks with genomic coordinates as human readable\"\"\"\n",
    "    if y:\n",
    "        ax.yaxis.set_major_formatter(bp_formatter)\n",
    "    if x:\n",
    "        ax.xaxis.set_major_formatter(bp_formatter)\n",
    "        ax.xaxis.tick_bottom()\n",
    "    if rotate:\n",
    "        ax.tick_params(axis='x',rotation=45)\n",
    "        \n",
    "# Functions to help with plotting\n",
    "def pcolormesh_45deg(ax, matrix_c, start=0, resolution=1, *args, **kwargs):\n",
    "    start_pos_vector = [start+resolution*i for i in range(len(matrix_c)+1)]\n",
    "    import itertools\n",
    "    n = matrix_c.shape[0]\n",
    "    t = np.array([[1, 0.5], [-1, 0.5]])\n",
    "    matrix_a = np.dot(np.array([(i[1], i[0])\n",
    "                                for i in itertools.product(start_pos_vector[::-1],\n",
    "                                                           start_pos_vector)]), t)\n",
    "    x = matrix_a[:, 1].reshape(n + 1, n + 1)\n",
    "    y = matrix_a[:, 0].reshape(n + 1, n + 1)\n",
    "    im = ax.pcolormesh(x, y, np.flipud(matrix_c), *args, **kwargs)\n",
    "    im.set_rasterized(True)\n",
    "    return im\n",
    "\n",
    "from matplotlib.ticker import EngFormatter\n",
    "bp_formatter = EngFormatter('b')\n",
    "def format_ticks(ax, x=True, y=True, rotate=True):\n",
    "    if y:\n",
    "        ax.yaxis.set_major_formatter(bp_formatter)\n",
    "    if x:\n",
    "        ax.xaxis.set_major_formatter(bp_formatter)\n",
    "        ax.xaxis.tick_bottom()\n",
    "    if rotate:\n",
    "        ax.tick_params(axis='x',rotation=45)\n",
    "        \n",
    "# create a functions that would return a series of rectangles around called dots\n",
    "# in a specific region, and exposing importnat plotting parameters\n",
    "def rectangles_around_dots(dots_df, region, loc=\"upper\", lw=1, ec=\"cyan\", fc=\"none\"):\n",
    "    \"\"\"\n",
    "    yield a series of rectangles around called dots in a given region\n",
    "    \"\"\"\n",
    "    # select dots from the region:\n",
    "    df_reg = bioframe.select(\n",
    "        bioframe.select(dots_df, region, cols=(\"chrom1\",\"start1\",\"end1\")),\n",
    "        region,\n",
    "        cols=(\"chrom2\",\"start2\",\"end2\"),\n",
    "    )\n",
    "    rectangle_kwargs = dict(lw=lw, ec=ec, fc=fc)\n",
    "    # draw rectangular \"boxes\" around pixels called as dots in the \"region\":\n",
    "    for s1, s2, e1, e2 in df_reg[[\"start1\", \"start2\", \"end1\", \"end2\"]].itertuples(index=False):\n",
    "        width1 = e1 - s1\n",
    "        width2 = e2 - s2\n",
    "        if loc == \"upper\":\n",
    "            yield patches.Rectangle((s2, s1), width2, width1, **rectangle_kwargs)\n",
    "        elif loc == \"lower\":\n",
    "            yield patches.Rectangle((s1, s2), width1, width2, **rectangle_kwargs)\n",
    "        else:\n",
    "            raise ValueError(\"loc has to be uppper or lower\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "60e71210",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0hr\n",
      "0hr: completed\n",
      "\n",
      "24hrs\n",
      "24hrs: completed\n",
      "\n",
      "72hrs\n",
      "72hrs: completed\n",
      "\n",
      "WTCI\n",
      "WTCI: completed\n",
      "\n",
      "KOCI\n",
      "KOCI: completed\n",
      "\n",
      "WTCIR\n",
      "WTCIR: completed\n",
      "\n",
      "KOCIR\n",
      "KOCIR: completed\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import cooltools.lib.plotting\n",
    "\n",
    "files = ['0hr', '24hrs', '72hrs', 'WTCI', 'KOCI', 'WTCIR', 'KOCIR']\n",
    "\n",
    "for file in files:\n",
    "    \n",
    "    print(file)\n",
    "    res_kb = 100\n",
    "    resname = f'{res_kb}kb'\n",
    "    resval = f'{res_kb}000'\n",
    "    \n",
    "    data_dir = '../cool/'\n",
    "    cool_file = data_dir + file + '.' + resname + '.mcool'\n",
    "    clr = cooler.Cooler(cool_file + '::/resolutions/' + resval)\n",
    "    resolution = clr.binsize\n",
    "    \n",
    "    # Load Genome Fasta\n",
    "    bins = clr.bins()[:]\n",
    "    hg38_genome = bioframe.load_fasta('/media/hanjun/Hanjun_Lee_Book/Genome/hg38.chromosomal.XYM.fa');\n",
    "    gc_cov = bioframe.frac_gc(bins[['chrom', 'start', 'end']], hg38_genome)\n",
    "    gc_cov.to_csv('../data/hg38_gc_cov_100kb.tsv',index=False,sep='\\t')\n",
    "    \n",
    "    # Use bioframe to fetch the genomic features from the UCSC.\n",
    "    hg38_chromsizes = bioframe.fetch_chromsizes('hg38')\n",
    "    hg38_cens = bioframe.fetch_centromeres('hg38')\n",
    "    hg38_arms = bioframe.make_chromarms(hg38_chromsizes, hg38_cens)\n",
    "    \n",
    "    view_df = pd.DataFrame({\n",
    "        'chrom': clr.chromnames,\n",
    "        'start': 0,\n",
    "        'end': clr.chromsizes.values,\n",
    "        'name': clr.chromnames,\n",
    "    })\n",
    "    \n",
    "    # eigenvector calculation\n",
    "    cis_eigs = cooltools.eigs_cis(\n",
    "        clr,\n",
    "        gc_cov,\n",
    "        view_df=view_df,\n",
    "        n_eigs=3,\n",
    "    )\n",
    "    \n",
    "    eigenvector_track = cis_eigs[1][['chrom','start','end','E1']]\n",
    "    \n",
    "    eigenvector_track.to_csv('../compartment/' + file + '.' + resname + '.tsv', sep='\\t', header=True)\n",
    "    \n",
    "    # bigWig file generation\n",
    "    bioframe.to_bigwig(eigenvector_track[['chrom','start','end','E1']], hg38_chromsizes, '../compartment/' + file + '.' + resname + '.bigWig')\n",
    "    \n",
    "    with open('../compartment/' + file + '.' + resname + '.pkl', 'wb') as f:\n",
    "        pickle.dump([cis_eigs, eigenvector_track], f)\n",
    "    \n",
    "    print(file + ': completed\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c04f69e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0hr\n",
      "0hr: completed\n",
      "\n",
      "24hrs\n",
      "24hrs: completed\n",
      "\n",
      "72hrs\n",
      "72hrs: completed\n",
      "\n",
      "WTCI\n",
      "WTCI: completed\n",
      "\n",
      "KOCI\n",
      "KOCI: completed\n",
      "\n",
      "WTCIR\n",
      "WTCIR: completed\n",
      "\n",
      "KOCIR\n",
      "KOCIR: completed\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import saddleplot\n",
    "import cooltools.lib.plotting\n",
    "from matplotlib.colors import LogNorm\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "import warnings\n",
    "from cytoolz import merge\n",
    "\n",
    "files = ['0hr', '24hrs', '72hrs', 'WTCI', 'KOCI', 'WTCIR', 'KOCIR']\n",
    "\n",
    "for file in files:\n",
    "    \n",
    "    print(file)\n",
    "    res_kb = 100\n",
    "    resname = f'{res_kb}kb'\n",
    "    resval = f'{res_kb}000'\n",
    "    \n",
    "    data_dir = '../cool/'\n",
    "    cool_file = data_dir + file + '.' + resname + '.mcool'\n",
    "    clr = cooler.Cooler(cool_file + '::/resolutions/' + resval)\n",
    "    resolution = clr.binsize\n",
    "    \n",
    "    # Load Genome Fasta\n",
    "    bins = clr.bins()[:]\n",
    "    hg38_genome = bioframe.load_fasta('/media/hanjun/Hanjun_Lee_Book/Genome/hg38.chromosomal.XYM.fa');\n",
    "    gc_cov = bioframe.frac_gc(bins[['chrom', 'start', 'end']], hg38_genome)\n",
    "    gc_cov.to_csv('../data/hg38_gc_cov_100kb.tsv',index=False,sep='\\t')\n",
    "    \n",
    "    # Use bioframe to fetch the genomic features from the UCSC.\n",
    "    hg38_chromsizes = bioframe.fetch_chromsizes('hg38')\n",
    "    hg38_cens = bioframe.fetch_centromeres('hg38')\n",
    "    hg38_arms = bioframe.make_chromarms(hg38_chromsizes, hg38_cens)\n",
    "    \n",
    "    with open('../compartment/' + file + '.' + resname + '.pkl', 'rb') as f:\n",
    "        cis_eigs, eigenvector_track = pickle.load(f)\n",
    "    \n",
    "    view_df = pd.DataFrame({\n",
    "        'chrom': clr.chromnames,\n",
    "        'start': 0,\n",
    "        'end': clr.chromsizes.values,\n",
    "        'name': clr.chromnames,\n",
    "    })\n",
    "    \n",
    "    cvd = cooltools.expected_cis(\n",
    "        clr=clr,\n",
    "        view_df=view_df,\n",
    "    )\n",
    "    \n",
    "    Q_LO = 0.025 # ignore 2.5% of genomic bins with the lowest E1 values\n",
    "    Q_HI = 0.975 # ignore 2.5% of genomic bins with the highest E1 values\n",
    "    N_GROUPS = 38 # divide remaining 95% of the genome into 38 equisized groups, 2.5% each\n",
    "    \n",
    "    interaction_sum, interaction_count =  cooltools.saddle(\n",
    "        clr,\n",
    "        cvd,\n",
    "        eigenvector_track,\n",
    "        'cis',\n",
    "        n_bins=N_GROUPS,\n",
    "        qrange=(Q_LO,Q_HI),\n",
    "        view_df=view_df\n",
    "    )\n",
    "    \n",
    "    with open('../compartment/' + file + '.saddle.' + resname + '.pkl', 'wb') as f:\n",
    "        pickle.dump([interaction_sum, interaction_count, Q_LO, Q_HI, N_GROUPS, cvd], f)\n",
    "    \n",
    "    print(file + ': completed\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ebdf6530",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0hr\n",
      "cbar\n",
      "0hr: completed\n",
      "\n",
      "24hrs\n",
      "cbar\n",
      "24hrs: completed\n",
      "\n",
      "72hrs\n",
      "cbar\n",
      "72hrs: completed\n",
      "\n",
      "WTCI\n",
      "cbar\n",
      "WTCI: completed\n",
      "\n",
      "KOCI\n",
      "cbar\n",
      "KOCI: completed\n",
      "\n",
      "WTCIR\n",
      "cbar\n",
      "WTCIR: completed\n",
      "\n",
      "KOCIR\n",
      "cbar\n",
      "KOCIR: completed\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 500x500 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 500x500 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 500x500 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 500x500 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 500x500 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 500x500 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 500x500 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import warnings\n",
    "from cytoolz import merge\n",
    "\n",
    "def saddleplot(\n",
    "    track,\n",
    "    saddledata,\n",
    "    n_bins,\n",
    "    vrange=None,\n",
    "    qrange=(0.0, 1.0),\n",
    "    cmap=\"coolwarm\",\n",
    "    scale=\"log\",\n",
    "    vmin=0.5,\n",
    "    vmax=2,\n",
    "    color=None,\n",
    "    title=None,\n",
    "    xlabel=None,\n",
    "    ylabel=None,\n",
    "    clabel=None,\n",
    "    fig=None,\n",
    "    fig_kws=None,\n",
    "    heatmap_kws=None,\n",
    "    margin_kws=None,\n",
    "    cbar_kws=None,\n",
    "    subplot_spec=None,\n",
    "):\n",
    "    \"\"\"\n",
    "    Generate a saddle plot.\n",
    "    Parameters\n",
    "    ----------\n",
    "    track : pd.DataFrame\n",
    "        See cooltools.digitize() for details.\n",
    "    saddledata : 2D array-like\n",
    "        Saddle matrix produced by `make_saddle`. It will include 2 flanking\n",
    "        rows/columns for outlier signal values, thus the shape should be\n",
    "        `(n+2, n+2)`.\n",
    "    cmap : str or matplotlib colormap\n",
    "        Colormap to use for plotting the saddle heatmap\n",
    "    scale : str\n",
    "        Color scaling to use for plotting the saddle heatmap: log or linear\n",
    "    vmin, vmax : float\n",
    "        Value limits for coloring the saddle heatmap\n",
    "    color : matplotlib color value\n",
    "        Face color for margin bar plots\n",
    "    fig : matplotlib Figure, optional\n",
    "        Specified figure to plot on. A new figure is created if none is\n",
    "        provided.\n",
    "    fig_kws : dict, optional\n",
    "        Passed on to `plt.Figure()`\n",
    "    heatmap_kws : dict, optional\n",
    "        Passed on to `ax.imshow()`\n",
    "    margin_kws : dict, optional\n",
    "        Passed on to `ax.bar()` and `ax.barh()`\n",
    "    cbar_kws : dict, optional\n",
    "        Passed on to `plt.colorbar()`\n",
    "    subplot_spec : GridSpec object\n",
    "        Specify a subregion of a figure to using a GridSpec.\n",
    "    Returns\n",
    "    -------\n",
    "    Dictionary of axes objects.\n",
    "    \"\"\"\n",
    "\n",
    "#     warnings.warn(\n",
    "#         \"Generating a saddleplot will be deprecated in future versions, \"\n",
    "#         + \"please see https://github.com/open2c_examples for examples on how to plot saddles.\",\n",
    "#         DeprecationWarning,\n",
    "#     )\n",
    "\n",
    "    from matplotlib.gridspec import GridSpec, GridSpecFromSubplotSpec\n",
    "    from matplotlib.colors import Normalize, LogNorm\n",
    "    from matplotlib import ticker\n",
    "    import matplotlib.pyplot as plt\n",
    "\n",
    "    class MinOneMaxFormatter(ticker.LogFormatter):\n",
    "        def set_locs(self, locs=None):\n",
    "            self._sublabels = set([vmin % 10 * 10, vmax % 10, 1])\n",
    "\n",
    "        def __call__(self, x, pos=None):\n",
    "            if x not in [vmin, 1, vmax]:\n",
    "                return \"\"\n",
    "            else:\n",
    "                return \"{x:g}\".format(x=x)\n",
    "\n",
    "    track_value_col = track.columns[3]\n",
    "    track_values = track[track_value_col].values\n",
    "\n",
    "    digitized_track, binedges = cooltools.digitize(\n",
    "        track, n_bins, vrange=vrange, qrange=qrange\n",
    "    )\n",
    "    x = digitized_track[digitized_track.columns[3]].values.astype(int).copy()\n",
    "    x = x[(x > -1) & (x < len(binedges) + 1)]\n",
    "\n",
    "    # Old version\n",
    "    # hist = np.bincount(x, minlength=len(binedges) + 1)\n",
    "\n",
    "    groupmean = track[track.columns[3]].groupby(digitized_track[digitized_track.columns[3]]).mean()\n",
    "\n",
    "    if qrange is not None:\n",
    "        lo, hi = qrange\n",
    "        binedges = np.linspace(lo, hi, n_bins + 1)\n",
    "\n",
    "    # Barplot of mean values and saddledata are flanked by outlier bins\n",
    "    n = saddledata.shape[0]\n",
    "    X, Y = np.meshgrid(binedges, binedges)\n",
    "    C = saddledata\n",
    "    if (n - n_bins) == 2:\n",
    "        C = C[1:-1, 1:-1]\n",
    "        groupmean = groupmean[1:-1]\n",
    "\n",
    "    # Layout\n",
    "    if subplot_spec is not None:\n",
    "        GridSpec = partial(GridSpecFromSubplotSpec, subplot_spec=subplot_spec)\n",
    "    grid = {}\n",
    "    gs = GridSpec(\n",
    "        nrows=3,\n",
    "        ncols=3,\n",
    "        width_ratios=[0.2, 1, 0.1],\n",
    "        height_ratios=[0.2, 1, 0.1],\n",
    "        wspace=0.05,\n",
    "        hspace=0.05,\n",
    "    )\n",
    "\n",
    "    # Figure\n",
    "    if fig is None:\n",
    "        fig_kws_default = dict(figsize=(5, 5))\n",
    "        fig_kws = merge(fig_kws_default, fig_kws if fig_kws is not None else {})\n",
    "        fig = plt.figure(**fig_kws)\n",
    "\n",
    "    # Heatmap\n",
    "    if scale == \"log\":\n",
    "        norm = LogNorm(vmin=vmin, vmax=vmax)\n",
    "    elif scale == \"linear\":\n",
    "        norm = Normalize(vmin=vmin, vmax=vmax)\n",
    "    else:\n",
    "        raise ValueError(\"Only linear and log color scaling is supported\")\n",
    "\n",
    "    grid[\"ax_heatmap\"] = ax = plt.subplot(gs[4])\n",
    "    heatmap_kws_default = dict(cmap=\"coolwarm\", rasterized=True)\n",
    "    heatmap_kws = merge(\n",
    "        heatmap_kws_default, heatmap_kws if heatmap_kws is not None else {}\n",
    "    )\n",
    "    img = ax.pcolormesh(X, Y, C, norm=norm, **heatmap_kws)\n",
    "    plt.gca().yaxis.set_visible(False)\n",
    "\n",
    "    # Margins\n",
    "    margin_kws_default = dict(edgecolor=\"k\", facecolor=color, linewidth=1)\n",
    "    margin_kws = merge(margin_kws_default, margin_kws if margin_kws is not None else {})\n",
    "    # left margin hist\n",
    "    grid[\"ax_margin_y\"] = plt.subplot(gs[3], sharey=grid[\"ax_heatmap\"])\n",
    "\n",
    "    plt.barh(\n",
    "        binedges, height=1/len(binedges), width=groupmean, align=\"edge\", **margin_kws\n",
    "    )\n",
    "\n",
    "    plt.xlim(plt.xlim()[1], plt.xlim()[0])  # fliplr\n",
    "    plt.ylim(hi, lo)\n",
    "    plt.gca().spines[\"top\"].set_visible(False)\n",
    "    plt.gca().spines[\"bottom\"].set_visible(False)\n",
    "    plt.gca().spines[\"left\"].set_visible(False)\n",
    "    plt.gca().xaxis.set_visible(False)\n",
    "    # top margin hist\n",
    "    grid[\"ax_margin_x\"] = plt.subplot(gs[1], sharex=grid[\"ax_heatmap\"])\n",
    "\n",
    "    plt.bar(\n",
    "        binedges, width=1/len(binedges), height=groupmean, align=\"edge\", **margin_kws\n",
    "    )\n",
    "\n",
    "    plt.xlim(lo, hi)\n",
    "    # plt.ylim(plt.ylim())  # correct\n",
    "    plt.gca().spines[\"top\"].set_visible(False)\n",
    "    plt.gca().spines[\"right\"].set_visible(False)\n",
    "    plt.gca().spines[\"left\"].set_visible(False)\n",
    "    plt.gca().xaxis.set_visible(False)\n",
    "    plt.gca().yaxis.set_visible(False)\n",
    "\n",
    "#     # Colorbar\n",
    "    grid[\"ax_cbar\"] = plt.subplot(gs[5])\n",
    "    cbar_kws_default = dict(fraction=0.8, label=clabel or \"\")\n",
    "    cbar_kws = merge(cbar_kws_default, cbar_kws if cbar_kws is not None else {})\n",
    "    if scale == \"linear\" and vmin is not None and vmax is not None:\n",
    "        grid[\"ax_cbar\"] = cb = plt.colorbar(img, **cbar_kws)\n",
    "        # cb.set_ticks(np.arange(vmin, vmax + 0.001, 0.5))\n",
    "        # # do linspace between vmin and vmax of 5 segments and trunc to 1 decimal:\n",
    "        decimal = 10\n",
    "        nsegments = 5\n",
    "        cd_ticks = np.trunc(np.linspace(vmin, vmax, nsegments) * decimal) / decimal\n",
    "        cb.set_ticks(cd_ticks)\n",
    "    else:\n",
    "        print('cbar')\n",
    "\n",
    "        cb = plt.colorbar(img, format=MinOneMaxFormatter(), cax=grid[\"ax_cbar\"], **cbar_kws)\n",
    "        cb.ax.yaxis.set_minor_formatter(MinOneMaxFormatter())\n",
    "\n",
    "    # extra settings\n",
    "    grid[\"ax_heatmap\"].set_xlim(lo, hi)\n",
    "    grid[\"ax_heatmap\"].set_ylim(hi, lo)\n",
    "    grid['ax_heatmap'].grid(False)\n",
    "    if title is not None:\n",
    "        grid[\"ax_margin_x\"].set_title(title)\n",
    "    if xlabel is not None:\n",
    "        grid[\"ax_heatmap\"].set_xlabel(xlabel)\n",
    "    if ylabel is not None:\n",
    "        grid[\"ax_margin_y\"].set_ylabel(ylabel)\n",
    "\n",
    "    return grid\n",
    "\n",
    "import cooltools.lib.plotting\n",
    "from matplotlib.colors import LogNorm\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "import warnings\n",
    "from cytoolz import merge\n",
    "\n",
    "files = ['0hr', '24hrs', '72hrs', 'WTCI', 'KOCI', 'WTCIR', 'KOCIR']\n",
    "\n",
    "for file in files:\n",
    "    \n",
    "    print(file)\n",
    "    res_kb = 100\n",
    "    resname = f'{res_kb}kb'\n",
    "    resval = f'{res_kb}000'\n",
    "    \n",
    "    data_dir = '../cool/'\n",
    "    cool_file = data_dir + file + '.' + resname + '.mcool'\n",
    "    clr = cooler.Cooler(cool_file + '::/resolutions/' + resval)\n",
    "    resolution = clr.binsize\n",
    "    \n",
    "    # Load Genome Fasta\n",
    "    bins = clr.bins()[:]\n",
    "    hg38_genome = bioframe.load_fasta('/media/hanjun/Hanjun_Lee_Book/Genome/hg38.chromosomal.XYM.fa');\n",
    "    gc_cov = bioframe.frac_gc(bins[['chrom', 'start', 'end']], hg38_genome)\n",
    "    gc_cov.to_csv('../data/hg38_gc_cov_100kb.tsv',index=False,sep='\\t')\n",
    "    \n",
    "    # Use bioframe to fetch the genomic features from the UCSC.\n",
    "    hg38_chromsizes = bioframe.fetch_chromsizes('hg38')\n",
    "    hg38_cens = bioframe.fetch_centromeres('hg38')\n",
    "    hg38_arms = bioframe.make_chromarms(hg38_chromsizes, hg38_cens)\n",
    "    \n",
    "    with open('../compartment/' + file + '.' + resname + '.pkl', 'rb') as f:\n",
    "        cis_eigs, eigenvector_track = pickle.load(f)\n",
    "    \n",
    "    with open('../compartment/' + file + '.saddle.' + resname + '.pkl', 'rb') as f:\n",
    "        interaction_sum, interaction_count, Q_LO, Q_HI, N_GROUPS, cvd = pickle.load(f)\n",
    "    \n",
    "    saddleplot(eigenvector_track,\n",
    "           interaction_sum/interaction_count,\n",
    "           N_GROUPS,\n",
    "           qrange=(Q_LO,Q_HI),\n",
    "           cbar_kws={'label':'average observed/expected contact frequency'}\n",
    "    );\n",
    "    \n",
    "    plt.savefig('../plot/' + file + '.saddle.' + resname + '.pdf')\n",
    "    plt.clf()\n",
    "    \n",
    "    print(file + ': completed\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "50586413",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0hr\n",
      "cbar\n",
      "0hr: completed\n",
      "\n",
      "24hrs\n",
      "cbar\n",
      "24hrs: completed\n",
      "\n",
      "72hrs\n",
      "cbar\n",
      "72hrs: completed\n",
      "\n",
      "WTCI\n",
      "cbar\n",
      "WTCI: completed\n",
      "\n",
      "KOCI\n",
      "cbar\n",
      "KOCI: completed\n",
      "\n",
      "WTCIR\n",
      "cbar\n",
      "WTCIR: completed\n",
      "\n",
      "KOCIR\n",
      "cbar\n",
      "KOCIR: completed\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 500x500 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 500x500 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 500x500 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 500x500 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 500x500 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 500x500 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 500x500 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import warnings\n",
    "from cytoolz import merge\n",
    "\n",
    "def saddleplot(\n",
    "    track,\n",
    "    saddledata,\n",
    "    n_bins,\n",
    "    vrange=None,\n",
    "    qrange=(0.0, 1.0),\n",
    "    cmap=\"coolwarm\",\n",
    "    scale=\"log\",\n",
    "    vmin=0.5,\n",
    "    vmax=2,\n",
    "    color=None,\n",
    "    title=None,\n",
    "    xlabel=None,\n",
    "    ylabel=None,\n",
    "    clabel=None,\n",
    "    fig=None,\n",
    "    fig_kws=None,\n",
    "    heatmap_kws=None,\n",
    "    margin_kws=None,\n",
    "    cbar_kws=None,\n",
    "    subplot_spec=None,\n",
    "):\n",
    "    \"\"\"\n",
    "    Generate a saddle plot.\n",
    "    Parameters\n",
    "    ----------\n",
    "    track : pd.DataFrame\n",
    "        See cooltools.digitize() for details.\n",
    "    saddledata : 2D array-like\n",
    "        Saddle matrix produced by `make_saddle`. It will include 2 flanking\n",
    "        rows/columns for outlier signal values, thus the shape should be\n",
    "        `(n+2, n+2)`.\n",
    "    cmap : str or matplotlib colormap\n",
    "        Colormap to use for plotting the saddle heatmap\n",
    "    scale : str\n",
    "        Color scaling to use for plotting the saddle heatmap: log or linear\n",
    "    vmin, vmax : float\n",
    "        Value limits for coloring the saddle heatmap\n",
    "    color : matplotlib color value\n",
    "        Face color for margin bar plots\n",
    "    fig : matplotlib Figure, optional\n",
    "        Specified figure to plot on. A new figure is created if none is\n",
    "        provided.\n",
    "    fig_kws : dict, optional\n",
    "        Passed on to `plt.Figure()`\n",
    "    heatmap_kws : dict, optional\n",
    "        Passed on to `ax.imshow()`\n",
    "    margin_kws : dict, optional\n",
    "        Passed on to `ax.bar()` and `ax.barh()`\n",
    "    cbar_kws : dict, optional\n",
    "        Passed on to `plt.colorbar()`\n",
    "    subplot_spec : GridSpec object\n",
    "        Specify a subregion of a figure to using a GridSpec.\n",
    "    Returns\n",
    "    -------\n",
    "    Dictionary of axes objects.\n",
    "    \"\"\"\n",
    "\n",
    "#     warnings.warn(\n",
    "#         \"Generating a saddleplot will be deprecated in future versions, \"\n",
    "#         + \"please see https://github.com/open2c_examples for examples on how to plot saddles.\",\n",
    "#         DeprecationWarning,\n",
    "#     )\n",
    "\n",
    "    from matplotlib.gridspec import GridSpec, GridSpecFromSubplotSpec\n",
    "    from matplotlib.colors import Normalize, LogNorm\n",
    "    from matplotlib import ticker\n",
    "    import matplotlib.pyplot as plt\n",
    "\n",
    "    class MinOneMaxFormatter(ticker.LogFormatter):\n",
    "        def set_locs(self, locs=None):\n",
    "            self._sublabels = set([vmin % 10 * 10, vmax % 10, 1])\n",
    "\n",
    "        def __call__(self, x, pos=None):\n",
    "            if x not in [vmin, 1, vmax]:\n",
    "                return \"\"\n",
    "            else:\n",
    "                return \"{x:g}\".format(x=x)\n",
    "\n",
    "    track_value_col = track.columns[3]\n",
    "    track_values = track[track_value_col].values\n",
    "\n",
    "    digitized_track, binedges = cooltools.digitize(\n",
    "        track, n_bins, vrange=vrange, qrange=qrange\n",
    "    )\n",
    "    x = digitized_track[digitized_track.columns[3]].values.astype(int).copy()\n",
    "    x = x[(x > -1) & (x < len(binedges) + 1)]\n",
    "\n",
    "    # Old version\n",
    "    # hist = np.bincount(x, minlength=len(binedges) + 1)\n",
    "\n",
    "    groupmean = track[track.columns[3]].groupby(digitized_track[digitized_track.columns[3]]).mean()\n",
    "\n",
    "    if qrange is not None:\n",
    "        lo, hi = qrange\n",
    "        binedges = np.linspace(lo, hi, n_bins + 1)\n",
    "\n",
    "    # Barplot of mean values and saddledata are flanked by outlier bins\n",
    "    n = saddledata.shape[0]\n",
    "    X, Y = np.meshgrid(binedges, binedges)\n",
    "    C = saddledata\n",
    "    if (n - n_bins) == 2:\n",
    "        C = C[1:-1, 1:-1]\n",
    "        groupmean = groupmean[1:-1]\n",
    "\n",
    "    # Layout\n",
    "    if subplot_spec is not None:\n",
    "        GridSpec = partial(GridSpecFromSubplotSpec, subplot_spec=subplot_spec)\n",
    "    grid = {}\n",
    "    gs = GridSpec(\n",
    "        nrows=3,\n",
    "        ncols=3,\n",
    "        width_ratios=[0.2, 1, 0.1],\n",
    "        height_ratios=[0.2, 1, 0.1],\n",
    "        wspace=0.05,\n",
    "        hspace=0.05,\n",
    "    )\n",
    "\n",
    "    # Figure\n",
    "    if fig is None:\n",
    "        fig_kws_default = dict(figsize=(5, 5))\n",
    "        fig_kws = merge(fig_kws_default, fig_kws if fig_kws is not None else {})\n",
    "        fig = plt.figure(**fig_kws)\n",
    "\n",
    "    # Heatmap\n",
    "    if scale == \"log\":\n",
    "        norm = LogNorm(vmin=vmin, vmax=vmax)\n",
    "    elif scale == \"linear\":\n",
    "        norm = Normalize(vmin=vmin, vmax=vmax)\n",
    "    else:\n",
    "        raise ValueError(\"Only linear and log color scaling is supported\")\n",
    "\n",
    "    grid[\"ax_heatmap\"] = ax = plt.subplot(gs[4])\n",
    "    heatmap_kws_default = dict(cmap=\"coolwarm\", rasterized=True)\n",
    "    heatmap_kws = merge(\n",
    "        heatmap_kws_default, heatmap_kws if heatmap_kws is not None else {}\n",
    "    )\n",
    "    img = ax.pcolormesh(X, Y, C, norm=norm, **heatmap_kws)\n",
    "    plt.gca().yaxis.set_visible(False)\n",
    "\n",
    "    # Margins\n",
    "    margin_kws_default = dict(edgecolor=\"k\", facecolor=color, linewidth=1)\n",
    "    margin_kws = merge(margin_kws_default, margin_kws if margin_kws is not None else {})\n",
    "    # left margin hist\n",
    "    grid[\"ax_margin_y\"] = plt.subplot(gs[3], sharey=grid[\"ax_heatmap\"])\n",
    "\n",
    "    plt.barh(\n",
    "        binedges, height=1/len(binedges), width=groupmean, align=\"edge\", **margin_kws\n",
    "    )\n",
    "\n",
    "    plt.xlim(plt.xlim()[1], plt.xlim()[0])  # fliplr\n",
    "    plt.ylim(hi, lo)\n",
    "    plt.gca().spines[\"top\"].set_visible(False)\n",
    "    plt.gca().spines[\"bottom\"].set_visible(False)\n",
    "    plt.gca().spines[\"left\"].set_visible(False)\n",
    "    plt.gca().xaxis.set_visible(False)\n",
    "    # top margin hist\n",
    "    grid[\"ax_margin_x\"] = plt.subplot(gs[1], sharex=grid[\"ax_heatmap\"])\n",
    "\n",
    "    plt.bar(\n",
    "        binedges, width=1/len(binedges), height=groupmean, align=\"edge\", **margin_kws\n",
    "    )\n",
    "\n",
    "    plt.xlim(lo, hi)\n",
    "    # plt.ylim(plt.ylim())  # correct\n",
    "    plt.gca().spines[\"top\"].set_visible(False)\n",
    "    plt.gca().spines[\"right\"].set_visible(False)\n",
    "    plt.gca().spines[\"left\"].set_visible(False)\n",
    "    plt.gca().xaxis.set_visible(False)\n",
    "    plt.gca().yaxis.set_visible(False)\n",
    "\n",
    "#     # Colorbar\n",
    "    grid[\"ax_cbar\"] = plt.subplot(gs[5])\n",
    "    cbar_kws_default = dict(fraction=0.8, label=clabel or \"\")\n",
    "    cbar_kws = merge(cbar_kws_default, cbar_kws if cbar_kws is not None else {})\n",
    "    if scale == \"linear\" and vmin is not None and vmax is not None:\n",
    "        grid[\"ax_cbar\"] = cb = plt.colorbar(img, **cbar_kws)\n",
    "        # cb.set_ticks(np.arange(vmin, vmax + 0.001, 0.5))\n",
    "        # # do linspace between vmin and vmax of 5 segments and trunc to 1 decimal:\n",
    "        decimal = 10\n",
    "        nsegments = 5\n",
    "        cd_ticks = np.trunc(np.linspace(vmin, vmax, nsegments) * decimal) / decimal\n",
    "        cb.set_ticks(cd_ticks)\n",
    "    else:\n",
    "        print('cbar')\n",
    "\n",
    "        cb = plt.colorbar(img, format=MinOneMaxFormatter(), cax=grid[\"ax_cbar\"], **cbar_kws)\n",
    "        cb.ax.yaxis.set_minor_formatter(MinOneMaxFormatter())\n",
    "\n",
    "    # extra settings\n",
    "    grid[\"ax_heatmap\"].set_xlim(lo, hi)\n",
    "    grid[\"ax_heatmap\"].set_ylim(hi, lo)\n",
    "    grid['ax_heatmap'].grid(False)\n",
    "    if title is not None:\n",
    "        grid[\"ax_margin_x\"].set_title(title)\n",
    "    if xlabel is not None:\n",
    "        grid[\"ax_heatmap\"].set_xlabel(xlabel)\n",
    "    if ylabel is not None:\n",
    "        grid[\"ax_margin_y\"].set_ylabel(ylabel)\n",
    "\n",
    "    return grid\n",
    "\n",
    "import cooltools.lib.plotting\n",
    "from matplotlib.colors import LogNorm\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "import warnings\n",
    "from cytoolz import merge\n",
    "\n",
    "files = ['0hr', '24hrs', '72hrs', 'WTCI', 'KOCI', 'WTCIR', 'KOCIR']\n",
    "\n",
    "for file in files:\n",
    "    \n",
    "    print(file)\n",
    "    res_kb = 100\n",
    "    resname = f'{res_kb}kb'\n",
    "    resval = f'{res_kb}000'\n",
    "    \n",
    "    data_dir = '../cool/'\n",
    "    cool_file = data_dir + file + '.' + resname + '.mcool'\n",
    "    clr = cooler.Cooler(cool_file + '::/resolutions/' + resval)\n",
    "    resolution = clr.binsize\n",
    "    \n",
    "    # Load Genome Fasta\n",
    "    bins = clr.bins()[:]\n",
    "    hg38_genome = bioframe.load_fasta('/media/hanjun/Hanjun_Lee_Book/Genome/hg38.chromosomal.XYM.fa');\n",
    "    gc_cov = bioframe.frac_gc(bins[['chrom', 'start', 'end']], hg38_genome)\n",
    "    gc_cov.to_csv('../data/hg38_gc_cov_100kb.tsv',index=False,sep='\\t')\n",
    "    \n",
    "    # Use bioframe to fetch the genomic features from the UCSC.\n",
    "    hg38_chromsizes = bioframe.fetch_chromsizes('hg38')\n",
    "    hg38_cens = bioframe.fetch_centromeres('hg38')\n",
    "    hg38_arms = bioframe.make_chromarms(hg38_chromsizes, hg38_cens)\n",
    "    \n",
    "    with open('../compartment/WTCI.' + resname + '.pkl', 'rb') as f:\n",
    "        cis_eigs, eigenvector_track = pickle.load(f)\n",
    "    \n",
    "    with open('../compartment/' + file + '.saddle.' + resname + '.pkl', 'rb') as f:\n",
    "        interaction_sum, interaction_count, Q_LO, Q_HI, N_GROUPS, cvd = pickle.load(f)\n",
    "    \n",
    "    saddleplot(eigenvector_track,\n",
    "           interaction_sum/interaction_count,\n",
    "           N_GROUPS,\n",
    "           qrange=(Q_LO,Q_HI),\n",
    "           cbar_kws={'label':'average observed/expected contact frequency'}\n",
    "    );\n",
    "    \n",
    "    plt.savefig('../plot/' + file + '.saddle_at_fixed_grid.' + resname + '.pdf')\n",
    "    plt.clf()\n",
    "    \n",
    "    print(file + ': completed\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38019666",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
